nifi-nar-bundles/nifi-cdc/nifi-cdc-postgresql-bundle/nifi-cdc-postgresql-processors/src/main/java/org/apache/nifi/cdc/postgresql/event/Decoder.java

```diff
@@ -0,0 +1,532 @@
+     *                                      Epoch Dates
+     * @throws UnsupportedEncodingException
+     *                                      if decodes fails to convert bytes
+     *                                      to String
+     */
+    public HashMap<String, Object> decodeLogicalReplicationBuffer(ByteBuffer buffer, boolean includeBeginCommit,
+            boolean includeAllMetadata) throws ParseException, UnsupportedEncodingException {
+
+        HashMap<String, Object> message = new HashMap<String, Object>();
+
+        /* (Byte1) Identifies the message type. */
+        char msgType = (char) buffer.get(0);
+        int position = 1;
+
+        switch (msgType) {
+            /* Identifies the message as a BEGIN message. */
+            case 'B':
+
+                if (includeBeginCommit || includeAllMetadata) {
+                    message.put("type", "begin");
+
+                    /* (Int64) The final LSN of the transaction. */
+                    message.put("xLSNFinal", buffer.getLong(1));
+
+                    /*
+                     * (Int64) Commit timestamp of the transaction. The value is in number of
+                     * microseconds since PostgreSQL epoch (2000-01-01).
+                     */
+                    message.put("xCommitTime",
+                            getFormattedPostgreSQLEpochDate(buffer.getLong(9)));
+
+                    /* (Int32) Xid of the transaction. */
+                    message.put("xid", buffer.getInt(17));
+                }
+                return message;
+
+            /* Identifies the message as a COMMIT message. */
+            case 'C':
+
+                if (includeBeginCommit || includeAllMetadata) {
+                    message.put("type", "commit");
+
+                    /* (Byte1) Flags; currently unused (must be 0). */
+                    message.put("flags", buffer.get(1));
+
+                    /* (Int64) The LSN of the commit. */
+                    message.put("commitLSN", buffer.getLong(2));
+
+                    /* (Int64) The end LSN of the transaction. */
+                    message.put("xLSNEnd", buffer.getLong(10));
+
+                    /*
+                     * (Int64) Commit timestamp of the transaction. The value is in number of
+                     * microseconds since PostgreSQL epoch (2000-01-01).
+                     */
+                    message.put("xCommitTime", getFormattedPostgreSQLEpochDate(
+                            buffer.getLong(18)));
+                }
+                return message;
+
+            /* Identifies the message as an ORIGIN message. */
+            case 'O':
+
+                if (includeAllMetadata) {
+                    message.put("type", "origin");
+
+                    /* (Int64) The LSN of the commit on the origin server. */
+                    message.put("originLSN", buffer.getLong(1));
+
+                    buffer.position(9);
+                    byte[] bytes_O = new byte[buffer.remaining()];
```

Comment:
Use of the underscore character should be avoided in variable names.
```suggestion
                    byte[] origin = new byte[buffer.remaining()];
```